modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import memoize from '../../../src/main/resources/META-INF/resources/liferay/util/memoize';
+
+describe('memoize', () => {
+	it('invokes the provided function with the provided arguments', () => {
+		const mockFn = jest.fn((object, number, array, string) => [
+			object,
+			number,
+			array,
+			string,
+		]);
+
+		const memoizedMockFn = memoize(mockFn);
+
+		expect(memoizedMockFn({3: 4}, 4, ['foo', 'bar'], 'baz')).toEqual([
+			{3: 4},
+			4,
+			['foo', 'bar'],
+			'baz',
+		]);
+	});
+
+	it("caches the argument if it isn't already cached", () => {
+		const mockFn = jest.fn((object, number, array, string) => [
+			object,
+			number,
+			array,
+			string,
+		]);
+
+		const memoizedMockFn = memoize(mockFn);
+
+		memoizedMockFn({3: 4}, 4, ['foo', 'bar'], 'baz');
+
+		expect(memoizedMockFn.getCache().size).toBe(1);
+	});
+
+	it("throws an error if the provided argument isn't a function", () => {
+		expect(() => memoize({foo: 42})).toThrow(TypeError);
+	});
```

Comment:
Can we add some more functional type tests here to test the function of memoize? For example, 

- Ensuring that the provided function is only called once, even if the memo is invoked twice
- Ensuring the value is equal between two calls of the memo'd function
- Ensure that the same object is returned if a memo is called twice with same args

For example
```js
function retObj(a,b) {
    return {a,b}
}

retObj('foo', 'bar') !== retObj('foo', 'bar') // it would create new object each time

memoizedRetObj('foo', 'bar') === memoizedRetObj('foo', 'bar') // should return the same object
```

I'd also grab some of the tests from the lodash unit tests and add them here to make sure we aren't missing any edge cases

Reply:
Great suggestions, will update next week, thank you @bryceosterhaus